relayer/src/link/relay_path.rs

```diff
@@ -1413,87 +1575,135 @@ impl<ChainA: ChainHandle, ChainB: ChainHandle> RelayPath<ChainA, ChainB> {
+        odata: OperationalData,
+        chain_time: &ChainTime,
+        max_expected_time_per_block: &MaxBlockTime,
+        latest_height: &LatestHeight,
+    ) -> Result<OperationalData, LinkError>
+    where
+        ChainTime: Fn() -> Result<Instant, LinkError>,
+        MaxBlockTime: Fn() -> Result<Duration, LinkError>,
+        LatestHeight: Fn() -> Result<Height, LinkError>,
+    {
+        let (time_left, blocks_left) =
+            odata.conn_delay_remaining(chain_time, max_expected_time_per_block, latest_height)?;
+
+        match (time_left, blocks_left) {
+            (Duration::ZERO, 0) => {
+                info!(
                     "ready to fetch a scheduled op. data with batch of size {} targeting {}",
                     odata.batch.len(),
                     odata.target,
-                ),
-                Some(delay_left) => {
-                    info!(
-                        "waiting ({:?} left) for a scheduled op. data with batch of size {} targeting {}",
-                        delay_left,
-                        odata.batch.len(),
-                        odata.target,
-                    );
+                );
+                Ok(odata)
+            }
+            (Duration::ZERO, blocks_left) => {
+                info!(
+                    "waiting ({:?} blocks left) for a scheduled op. data with batch of size {} targeting {}",
+                    blocks_left,
+                    odata.batch.len(),
+                    odata.target,
+                );
 
-                    // Wait until the delay period passes
-                    thread::sleep(delay_left);
-                }
+                let blocks_left: u32 = blocks_left.try_into().expect("blocks_left > u32::MAX");
+
+                // Wait until the delay period passes
+                thread::sleep(blocks_left * max_expected_time_per_block()?);
+
+                Ok(odata)
             }
+            (time_left, _) => {
+                info!(
+                    "waiting ({:?} left) for a scheduled op. data with batch of size {} targeting {}",
+                    time_left,
+                    odata.batch.len(),
+                    odata.target,
+                );
+
+                // Wait until the delay period passes
+                thread::sleep(time_left);
 
-            Some(odata)
+                // `blocks_left` maybe non-zero, so recurse to recheck that all delays are handled.
+                Self::wait_for_conn_delay(
+                    odata,
+                    chain_time,
+                    max_expected_time_per_block,
+                    latest_height,
+                )
+            }
+        }
+    }
+
+    /// Fetches an operational data that has fulfilled its predefined delay period. May _block_
+    /// waiting for the delay period to pass.
+    /// Returns `Ok(None)` if there is no operational data scheduled.
+    pub(crate) fn fetch_scheduled_operational_data(
```

Comment:
I think this is a valid concern, although the code in question predates this PR. I just tried to extract this code (i.e. `fetch_scheduled_operational_data()`, `build_and_send_ack_packet_messages()`, etc.) into `relayer-cli/src/commands.rs` but it seems like it requires a lot of visibility changes which include making the `relay_path` mod public and exposing relayer internals like `SyncSender`, so I am a little reluctant to implement this. This would have been easier if there was a `pub(workspace)`.

Reply:
Alternatively can we move them into separate files in the same crate, i.e. `relayer/src/link/relay_path/cli.rs`? It is allowed to have multiple `impl RelayPath {}` blocks within the same crate.

Reply:
Nice idea! https://github.com/informalsystems/ibc-rs/pull/1917/commits/e0583586d24dfca4e2da9aeb5ced05d653843ecf